File: packages/toon/src/encode/encoders.ts
import type { Depth, FieldNode, JsonArray, JsonObject, JsonValue, ResolvedEncodeOptions } from '../types.ts'
import type { EncodablePrimitive } from './raw-string.ts'
import { LIST_ITEM_MARKER, LIST_ITEM_PREFIX } from '../constants.ts'
import { isArrayOfArrays, isArrayOfObjects, isArrayOfPrimitives, isEmptyObject, isEncodablePrimitive, isJsonArray, isJsonObject } from './normalize.ts'
import { encodeAndJoinPrimitives, encodeKey, encodePrimitive, formatHeader } from './primitives.ts'

// #region Encode normalized JsonValue

export function* encodeJsonValue(value: JsonValue, options: ResolvedEncodeOptions, depth: Depth): Generator<string> {
  if (isEncodablePrimitive(value)) {
    // Primitives at root level are returned as a single line
    const encodedPrimitive = encodePrimitive(value, options.delimiter)

    if (encodedPrimitive !== '')
      yield encodedPrimitive

    return
  }

  if (isJsonArray(value)) {
    yield* encodeArrayLines(undefined, value, depth, options)
  }
  else if (isJsonObject(value)) {
    // A keyed-eligible root object uses the keyless keyed form
    const keyedFields = extractKeyedFields(value)
    if (keyedFields) {
      yield* encodeKeyedObjectLines(undefined, value, keyedFields, depth, options)
      return
    }

    yield* encodeObjectLines(value, depth, options)
  }
}

// #endregion

// #region Object encoding

export function* encodeObjectLines(
  value: JsonObject,
  depth: Depth,
  options: ResolvedEncodeOptions,
): Generator<string> {
  for (const [key, val] of Object.entries(value)) {
    yield* encodeKeyValuePairLines(key, val, depth, options)
  }
}

export function* encodeKeyValuePairLines(
  key: string,
  value: JsonValue,
  depth: Depth,
  options: ResolvedEncodeOptions,
): Generator<string> {
  const encodedKey = encodeKey(key)

  if (isEncodablePrimitive(value)) {
    yield indentedLine(depth, `${encodedKey}: ${encodePrimitive(value, options.delimiter)}`, options.indent)
  }
  else if (isJsonArray(value)) {
    yield* encodeArrayLines(key, value, depth, options)
  }
  else if (isJsonObject(value)) {
    const keyedFields = extractKeyedFields(value)
    if (keyedFields) {
      yield* encodeKeyedObjectLines(key, value, keyedFields, depth, options)
      return
    }

    yield indentedLine(depth, `${encodedKey}:`, options.indent)
    if (!isEmptyObject(value)) {
      yield* encodeObjectLines(value, depth + 1, options)
    }
  }
}

// #endregion

// #region Keyed tabular objects

export function extractKeyedFields(value: JsonObject): FieldNode[] | undefined {
  const entryValues = Object.values(value)

  // At least two entries whose values are uniform non-empty objects
  if (entryValues.length < 2) {
    return
  }
  if (!entryValues.every(entryValue => isJsonObject(entryValue) && !isEmptyObject(entryValue))) {
    return
  }

  return extractTabularHeader(entryValues as JsonObject[])
}

function* encodeKeyedObjectLines(
  key: string | undefined,
  value: JsonObject,
  fields: readonly FieldNode[],
  depth: Depth,
  options: ResolvedEncodeOptions,
): Generator<string> {
  const entries = Object.entries(value)
  const header = formatHeader(entries.length, { key, fields, delimiter: options.delimiter, keyed: true })
  yield indentedLine(depth, header, options.indent)
  yield* encodeKeyedEntryRowsLines(entries, fields, depth + 1, options)
}

function* encodeKeyedEntryRowsLines(
  entries: readonly [string, JsonValue][],
  fields: readonly FieldNode[],
  depth: Depth,
  options: ResolvedEncodeOptions,
): Generator<string> {
  for (const [entryKey, entryValue] of entries) {
    const leaves: EncodablePrimitive[] = []
    collectLeafValues(entryValue as JsonObject, fields, leaves)
    yield indentedLine(depth, `${encodeKey(entryKey)}: ${encodeAndJoinPrimitives(leaves, options.delimiter)}`, options.indent)
  }
}

// #endregion

// #region Array encoding

export function* encodeArrayLines(
  key: string | undefined,
  value: JsonArray,
  depth: Depth,
  options: ResolvedEncodeOptions,
): Generator<string> {
  if (value.length === 0) {
    const line = key != null ? `${encodeKey(key)}: []` : '[]'
    yield indentedLine(depth, line, options.indent)
    return
  }

  // Primitive array
  if (isArrayOfPrimitives(value)) {
    const arrayLine = encodeInlineArrayLine(value, options.delimiter, key)
    yield indentedLine(depth, arrayLine, options.indent)
    return
  }

  // Array of arrays (all primitives)
  if (isArrayOfArrays(value)) {
    const allPrimitiveArrays = value.every(arr => isArrayOfPrimitives(arr))
    if (allPrimitiveArrays) {
      yield* encodeArrayOfArraysAsListItemsLines(key, value, depth, options)
      return
    }
  }

  // Array of objects
  if (isArrayOfObjects(value)) {
    const header = extractTabularHeader(value)
    if (header) {
      yield* encodeArrayOfObjectsAsTabularLines(key, value, header, depth, options)
    }
    else {
      yield* encodeMixedArrayAsListItemsLines(key, value, depth, options)
    }
    return
  }

  // Mixed array: fallback to expanded format
  yield* encodeMixedArrayAsListItemsLines(key, value, depth, options)
}

// #endregion

// #region Array of arrays (expanded format)

export function* encodeArrayOfArraysAsListItemsLines(
  prefix: string | undefined,
  values: readonly JsonArray[],
  depth: Depth,
  options: ResolvedEncodeOptions,
): Generator<string> {
  const header = formatHeader(values.length, { key: prefix, delimiter: options.delimiter })
  yield indentedLine(depth, header, options.indent)

  for (const arr of values) {
    if (isArrayOfPrimitives(arr)) {
      const arrayLine = encodeInlineArrayLine(arr, options.delimiter)
      yield indentedListItem(depth + 1, arrayLine, options.indent)
    }
  }
}

export function encodeInlineArrayLine(values: readonly EncodablePrimitive[], delimiter: string, prefix?: string): string {
  const header = formatHeader(values.length, { key: prefix, delimiter })
  const joinedValue = encodeAndJoinPrimitives(values, delimiter)

  if (values.length === 0)
    return header

  return `${header} ${joinedValue}`
}

// #endregion

// #region Array of objects (tabular format)

export function* encodeArrayOfObjectsAsTabularLines(
  prefix: string | undefined,
  rows: readonly JsonObject[],
  header: readonly FieldNode[],
  depth: Depth,
  options: ResolvedEncodeOptions,
): Generator<string> {
  const formattedHeader = formatHeader(rows.length, { key: prefix, fields: header, delimiter: options.delimiter })
  yield indentedLine(depth, formattedHeader, options.indent)

  yield* writeTabularRowsLines(rows, header, depth + 1, options)
}

export function extractTabularHeader(rows: readonly JsonObject[]): FieldNode[] | undefined {
  if (rows.length === 0)
    return

  const firstKeys = Object.keys(rows[0]!)
  if (firstKeys.length === 0)
    return

  // All objects must have the same set of keys (order per object may vary)
  for (const row of rows) {
    if (Object.keys(row).length !== firstKeys.length) {
      return
    }
    for (const key of firstKeys) {
      if (!Object.hasOwn(row, key)) {
        return
      }
    }
  }

  const fieldNodes: FieldNode[] = []
  for (const key of firstKeys) {
    const fieldNode = classifyColumn(key, rows.map(row => row[key]!))
    if (!fieldNode) {
      return
    }
    fieldNodes.push(fieldNode)
  }

  return fieldNodes
}

function classifyColumn(name: string, values: readonly JsonValue[]): FieldNode | undefined {
  // Uniform-primitive column: a bare leaf field
  if (values.every(value => isEncodablePrimitive(value))) {
    return { name }
  }

  // Nested-uniform column: every value a non-empty object sharing one key
  // set whose sub-columns classify recursively
  if (!values.every(value => isJsonObject(value) && !isEmptyObject(value))) {
    return
  }

  const children = extractTabularHeader(values as JsonObject[])
  if (!children) {
    return
  }

  return { name, children }
}

function collectLeafValues(row: JsonObject, fields: readonly FieldNode[], leaves: EncodablePrimitive[]): void {
  for (const field of fields) {
    const value = row[field.name]
    if (field.children) {
      collectLeafValues(value as JsonObject, field.children, leaves)
    }
    else {
      leaves.push(value as EncodablePrimitive)
    }
  }
}

function* writeTabularRowsLines(
  rows: readonly JsonObject[],
  header: readonly FieldNode[],
  depth: Depth,
  options: ResolvedEncodeOptions,
): Generator<string> {
  for (const row of rows) {
    const leaves: EncodablePrimitive[] = []
    collectLeafValues(row, header, leaves)
    yield indentedLine(depth, encodeAndJoinPrimitives(leaves, options.delimiter), options.indent)
  }
}

// #endregion

// #region Array of objects (expanded format)

export function* encodeMixedArrayAsListItemsLines(
  prefix: string | undefined,
  items: readonly JsonValue[],
  depth: Depth,
  options: ResolvedEncodeOptions,
): Generator<string> {
  const header = formatHeader(items.length, { key: prefix, delimiter: options.delimiter })
  yield indentedLine(depth, header, options.indent)

  for (const item of items) {
    yield* encodeListItemValueLines(item, depth + 1, options)
  }
}

export function* encodeObjectAsListItemLines(
  obj: JsonObject,
  depth: Depth,
  options: ResolvedEncodeOptions,
): Generator<string> {
  if (isEmptyObject(obj)) {
    yield indentedLine(depth, LIST_ITEM_MARKER, options.indent)
    return
  }

  const entries = Object.entries(obj)
  const [firstKey, firstValue] = entries[0]!
  const restEntries = entries.slice(1)

  // Check if first field is a tabular array
  if (isJsonArray(firstValue) && isArrayOfObjects(firstValue)) {
    const header = extractTabularHeader(firstValue)
    if (header) {
      // Tabular array as first field
      const formattedHeader = formatHeader(firstValue.length, { key: firstKey, fields: header, delimiter: options.delimiter })
      yield indentedListItem(depth, formattedHeader, options.indent)
      yield* writeTabularRowsLines(firstValue, header, depth + 2, options)

      if (restEntries.length > 0) {
        const restObj: JsonObject = Object.fromEntries(restEntries)
        yield* encodeObjectLines(restObj, depth + 1, options)
      }
      return
    }
  }

  // Check if first field is a keyed tabular object: header on the hyphen
  // line, entry rows at depth +2, sibling fields at depth +1
  if (isJsonObject(firstValue)) {
    const keyedFields = extractKeyedFields(firstValue)
    if (keyedFields) {
      const keyedEntries = Object.entries(firstValue)
      const formattedHeader = formatHeader(keyedEntries.length, { key: firstKey, fields: keyedFields, delimiter: options.delimiter, keyed: true })
      yield indentedListItem(depth, formattedHeader, options.indent)
      yield* encodeKeyedEntryRowsLines(keyedEntries, keyedFields, depth + 2, options)

      if (restEntries.length > 0) {
        const restObj: JsonObject = Object.fromEntries(restEntries)
        yield* encodeObjectLines(restObj, depth + 1, options)
      }
      return
    }
  }

  const encodedKey = encodeKey(firstKey)

  if (isEncodablePrimitive(firstValue)) {
    // Primitive value: `- key: value`
    const encodedValue = encodePrimitive(firstValue, options.delimiter)
    yield indentedListItem(depth, `${encodedKey}: ${encodedValue}`, options.indent)
  }
  else if (isJsonArray(firstValue)) {
    if (firstValue.length === 0) {
      // Empty array: `- key: []`
      yield indentedListItem(depth, `${encodedKey}: []`, options.indent)
    }
    else if (isArrayOfPrimitives(firstValue)) {
      // Inline primitive array: `- key[N]: values`
      const arrayLine = encodeInlineArrayLine(firstValue, options.delimiter)
      yield indentedListItem(depth, `${encodedKey}${arrayLine}`, options.indent)
    }
    else {
      // Non-inline array: `- key[N]:` with items at depth + 2
      const header = formatHeader(firstValue.length, { delimiter: options.delimiter })
      yield indentedListItem(depth, `${encodedKey}${header}`, options.indent)

      for (const item of firstValue) {
        yield* encodeListItemValueLines(item, depth + 2, options)
      }
    }
  }
  else if (isJsonObject(firstValue)) {
    // Object value: `- key:` with fields at depth + 2
    yield indentedListItem(depth, `${encodedKey}:`, options.indent)
    if (!isEmptyObject(firstValue)) {
      yield* encodeObjectLines(firstValue, depth + 2, options)
    }
  }

  if (restEntries.length > 0) {
    const restObj: JsonObject = Object.fromEntries(restEntries)
    yield* encodeObjectLines(restObj, depth + 1, options)
  }
}

// #endregion

// #region List item encoding helpers

function* encodeListItemValueLines(
  value: JsonValue,
  depth: Depth,
  options: ResolvedEncodeOptions,
): Generator<string> {
  if (isEncodablePrimitive(value)) {
    yield indentedListItem(depth, encodePrimitive(value, options.delimiter), options.indent)
  }
  else if (isJsonArray(value)) {
    if (isArrayOfPrimitives(value)) {
      const arrayLine = encodeInlineArrayLine(value, options.delimiter)
      yield indentedListItem(depth, arrayLine, options.indent)
    }
    else {
      const header = formatHeader(value.length, { delimiter: options.delimiter })
      yield indentedListItem(depth, header, options.indent)
      for (const item of value) {
        yield* encodeListItemValueLines(item, depth + 1, options)
      }
    }
  }
  else if (isJsonObject(value)) {
    yield* encodeObjectAsListItemLines(value, depth, options)
  }
}

// #endregion

// #region Indentation helpers

function indentedLine(depth: Depth, content: string, indentSize: number): string {
  const indentation = ' '.repeat(indentSize * depth)
  return indentation + content
}

function indentedListItem(depth: Depth, content: string, indentSize: number): string {
  return indentedLine(depth, LIST_ITEM_PREFIX + content, indentSize)
}

// #endregion

File: packages/toon_julia/src/encode/encoders.jl
"""
Encodes a JsonValue into TOON format string.
"""
function encode_value(value::JsonValue, options::EncodeOptions)::String
    if is_json_primitive(value)
        return encode_primitive(value, options.delimiter)
    end
    
    writer = LineWriter(options.indent)
    
    if is_json_array(value)
        encode_array(nothing, value, writer, 0, options)
    elseif is_json_object(value)
        encode_object(value, writer, 0, options)
    end
    
    return string(writer)
end

"""
Encodes an object.
"""
function encode_object(
    value::JsonObject,
    writer::LineWriter,
    depth::Int,
    options::EncodeOptions,
    root_literal_keys::Union{Set{String}, Nothing} = nothing,
    path_prefix::Union{String, Nothing} = nothing,
    remaining_depth::Union{Int, Nothing} = nothing
)
    keys_list = sort(collect(keys(value)))
    
    # At root level (depth 0), collect all literal dotted keys for collision checking
    if depth == 0 && root_literal_keys === nothing
        root_literal_keys = Set([k for k in keys_list if occursin(".", k)])
    end
    
    effective_flatten_depth = something(remaining_depth, options.flatten_depth)
    
    # Iterate over sorted keys
    for key in keys_list
        val = value[key]
        encode_key_value_pair(key, val, writer, depth, options, keys_list, root_literal_keys, path_prefix, effective_flatten_depth)
    end
end

"""
Encodes a key-value pair.
"""
function encode_key_value_pair(
    key::String,
    value::JsonValue,
    writer::LineWriter,
    depth::Int,
    options::EncodeOptions,
    siblings::Union{Vector{String}, Nothing} = nothing,
    root_literal_keys::Union{Set{String}, Nothing} = nothing,
    path_prefix::Union{String, Nothing} = nothing,
    flatten_depth::Union{Int, Nothing} = nothing
)
    current_path = path_prefix !== nothing ? "$(path_prefix)$(DOT)$(key)" : key
    effective_flatten_depth = something(flatten_depth, options.flatten_depth)
    
    # Key folding is not implemented in this version (can be added later)
    # For now, use standard encoding
    
    encoded_key = encode_key(key)
    
    if is_json_primitive(value)
        push(writer, depth, "$(encoded_key): $(encode_primitive(value, options.delimiter))")
    elseif is_json_array(value)
        encode_array(key, value, writer, depth, options)
    elseif is_json_object(value)
        push(writer, depth, "$(encoded_key):")
        if !is_empty_object(value)
            encode_object(value, writer, depth + 1, options, root_literal_keys, current_path, effective_flatten_depth)
        end
    end
end

"""
Encodes an array.
"""
function encode_array(
    key::Union{String, Nothing},
    value::JsonArray,
    writer::LineWriter,
    depth::Int,
    options::EncodeOptions
)
    if isempty(value)
        header = format_header(0; key = key, delimiter = options.delimiter)
        push(writer, depth, header)
        return
    end
    
    # Primitive array
    if is_array_of_primitives(value)
        # Convert to Vector{JsonPrimitive} for type compatibility
        primitive_array = JsonPrimitive[item for item in value]
        array_line = encode_inline_array_line(primitive_array, options.delimiter, key)
        push(writer, depth, array_line)
        return
    end
    
    # Array of arrays (all primitives)
    if is_array_of_arrays(value)
        all_primitive_arrays = all(arr -> is_json_array(arr) && is_array_of_primitives(arr), value)
        if all_primitive_arrays
            encode_array_of_arrays_as_list_items(key, value, writer, depth, options)
            return
        end
    end
    
    # Array of objects
    if is_array_of_objects(value)
        header = extract_tabular_header(value)
        if header !== nothing
            encode_array_of_objects_as_tabular(key, value, header, writer, depth, options)
        else
            encode_mixed_array_as_list_items(key, value, writer, depth, options)
        end
        return
    end
    
    # Mixed array: fallback to expanded format
    encode_mixed_array_as_list_items(key, value, writer, depth, options)
end

"""
Encodes an array of arrays as list items.
"""
function encode_array_of_arrays_as_list_items(
    prefix::Union{String, Nothing},
    values::JsonArray,
    writer::LineWriter,
    depth::Int,
    options::EncodeOptions
)
    header = format_header(length(values); key = prefix, delimiter = options.delimiter)
    push(writer, depth, header)
    
    for arr in values
        if is_json_array(arr) && is_array_of_primitives(arr)
            # Convert to JsonPrimitive[] for type compatibility
            primitive_array = JsonPrimitive[item for item in arr]
            array_line = encode_inline_array_line(primitive_array, options.delimiter, nothing)
            push_list_item(writer, depth + 1, array_line)
        end
    end
end

"""
Encodes an inline array line.
"""
function encode_inline_array_line(
    values::Vector{JsonPrimitive},
    delimiter::Char,
    prefix::Union{String, Nothing} = nothing
)::String
    header = format_header(length(values); key = prefix, delimiter = delimiter)
    joined_value = encode_and_join_primitives(values, delimiter)
    if isempty(values)
        return header
    end
    return "$(header) $(joined_value)"
end

"""
Extracts tabular header from an array of objects.
"""
function extract_tabular_header(rows::JsonArray)::Union{Vector{String}, Nothing}
    if isempty(rows)
        return nothing
    end
    
    first_row = rows[1]
    if !is_json_object(first_row)
        return nothing
    end
    
    # Try to preserve order from JSON Object if available
    # JSON.jl's Object type preserves insertion order
    first_keys_raw = collect(keys(first_row))
    first_keys = [string(k) for k in first_keys_raw]
    if isempty(first_keys)
        return nothing
    end
    
    # Check if all rows have the same keys (order-independent check)
    sorted_keys = sort(first_keys)
    if is_tabular_array(rows, sorted_keys)
        # For deterministic output, sort alphabetically
        # Note: This may not match test expectations that preserve JSON order,
        # but Dict iteration order is non-deterministic in Julia
        return sorted_keys
    end
    
    return nothing
end

"""
Checks if an array of objects is tabular (all have same keys, all values are primitives).
"""
function is_tabular_array(rows::JsonArray, header::Vector{String})::Bool
    for row in rows
        if !is_json_object(row)
            return false
        end
        
        row_keys = [string(k) for k in keys(row)]
        
        # All objects must have the same number of keys
        if length(row_keys) != length(header)
            return false
        end
        
        # Check that all header keys exist in the row and all values are primitives
        for key in header
            if !haskey(row, key)
                return false
            end
            if !is_json_primitive(row[key])
                return false
            end
        end
    end
    
    return true
end

"""
Encodes an array of objects as tabular format.
"""
function encode_array_of_objects_as_tabular(
    prefix::Union{String, Nothing},
    rows::JsonArray,
    header::Vector{String},
    writer::LineWriter,
    depth::Int,
    options::EncodeOptions
)
    formatted_header = format_header(
        length(rows);
        key = prefix,
        fields = header,
        delimiter = options.delimiter
    )
    push(writer, depth, formatted_header)
    
    write_tabular_rows(rows, header, writer, depth + 1, options)
end

"""
Writes tabular rows.
"""
function write_tabular_rows(
    rows::JsonArray,
    header::Vector{String},
    writer::LineWriter,
    depth::Int,
    options::EncodeOptions
)
    for row in rows
        if !is_json_object(row)
            continue
        end
        values = JsonPrimitive[]
        for key in header
            val = row[key]
            if is_json_primitive(val)
                push!(values, val)
            end
        end
        joined_value = encode_and_join_primitives(values, options.delimiter)
        push(writer, depth, joined_value)
    end
end

"""
Encodes a mixed array as list items.
"""
function encode_mixed_array_as_list_items(
    prefix::Union{String, Nothing},
    items::JsonArray,
    writer::LineWriter,
    depth::Int,
    options::EncodeOptions
)
    header = format_header(length(items); key = prefix, delimiter = options.delimiter)
    push(writer, depth, header)
    
    for item in items
        encode_list_item_value(item, writer, depth + 1, options)
    end
end

"""
Encodes an object as a list item.
"""
function encode_object_as_list_item(
    obj::JsonObject,
    writer::LineWriter,
    depth::Int,
    options::EncodeOptions
)
    if is_empty_object(obj)
        push_list_item(writer, depth, "")
        return
    end
    
    # Sort keys: primitives first (alphabetically), then arrays (alphabetically), then objects (alphabetically)
    all_keys = collect(keys(obj))
    primitive_keys = [k for k in all_keys if is_json_primitive(obj[k])]
    array_keys = [k for k in all_keys if is_json_array(obj[k])]
    object_keys = [k for k in all_keys if is_json_object(obj[k])]
    
    sorted_keys = vcat(sort(primitive_keys), sort(array_keys), sort(object_keys))
    first_key = sorted_keys[1]
    first_value = obj[first_key]
    encoded_key = encode_key(first_key)
    
    if is_json_primitive(first_value)
        push_list_item(writer, depth, "$(encoded_key): $(encode_primitive(first_value, options.delimiter))")
    elseif is_json_array(first_value)
        if is_array_of_primitives(first_value)
            # Inline format for primitive arrays
            # Convert to JsonPrimitive[] for type compatibility
            primitive_array = JsonPrimitive[item for item in first_value]
            array_property_line = encode_inline_array_line(primitive_array, options.delimiter, first_key)
            push_list_item(writer, depth, array_property_line)
        elseif is_array_of_objects(first_value)
            # Check if array of objects can use tabular format
            tabular_header = extract_tabular_header(first_value)
            if tabular_header !== nothing
                # Tabular format for uniform arrays of objects
                formatted_header = format_header(
                    length(first_value);
                    key = first_key,
                    fields = tabular_header,
                    delimiter = options.delimiter
                )
                push_list_item(writer, depth, formatted_header)
                write_tabular_rows(first_value, tabular_header, writer, depth + 1, options)
            else
                # Fall back to list format for non-uniform arrays of objects
                push_list_item(writer, depth, "$(encoded_key)[$(length(first_value))]:")
                for item in first_value
                    encode_object_as_list_item(item, writer, depth + 1, options)
                end
            end
        else
            # Complex arrays on separate lines (array of arrays, etc.)
            push_list_item(writer, depth, "$(encoded_key)[$(length(first_value))]:")
            
            # Encode array contents at depth + 1
            for item in first_value
                encode_list_item_value(item, writer, depth + 1, options)
            end
        end
    elseif is_json_object(first_value)
        push_list_item(writer, depth, "$(encoded_key):")
        if !is_empty_object(first_value)
            encode_object(first_value, writer, depth + 2, options)
        end
    end
    
    # Remaining entries on indented lines (sorted)
    for i in 2:length(sorted_keys)
        key = sorted_keys[i]
        val = obj[key]
        encode_key_value_pair(key, val, writer, depth + 1, options)
    end
end

"""
Encodes a list item value.
"""
function encode_list_item_value(
    value::JsonValue,
    writer::LineWriter,
    depth::Int,
    options::EncodeOptions
)
    if is_json_primitive(value)
        push_list_item(writer, depth, encode_primitive(value, options.delimiter))
    elseif is_json_array(value) && is_array_of_primitives(value)
        # Convert to JsonPrimitive[] for type compatibility
        primitive_array = JsonPrimitive[item for item in value]
        array_line = encode_inline_array_line(primitive_array, options.delimiter, nothing)
        push_list_item(writer, depth, array_line)
    elseif is_json_object(value)
        encode_object_as_list_item(value, writer, depth, options)
    elseif is_json_array(value)
        # Handle non-primitive arrays
        header = format_header(length(value); delimiter = options.delimiter)
        push_list_item(writer, depth, header)
        for item in value
            encode_list_item_value(item, writer, depth + 1, options)
        end
    end
end

